File: dotnet/Nitro/Geography/Datum.cs
using Nitro.Geometry;


namespace Nitro.Geography
{
    public class Datum
    {
        public static Datum WGS84 = new Datum(6378137, 1.0 / 298.257223563, 0, 0, 0);

        protected double a;
        protected double b;
        protected double e;
        protected double e_;
        protected double f;
        protected double xShift;
        protected double yShift;
        protected double zShift;

        public Datum(double a, double f, double xShift, double yShift, double zShift)
        {
            this.a = a;
            this.f = f;
            this.b = a * (1 - f);
            this.xShift = xShift;
            this.yShift = yShift;
            this.zShift = zShift;

            e = Math.Sqrt((a * a - b * b) / (a * a));
            e_ = Math.Sqrt((a * a - b * b) / (b * b));
        }

        public Vector3 ToCarthesian(Coordinate coordinate)
        {
            double radLat = (Math.PI / 180) * coordinate.Latitude;
            double radLon = (Math.PI / 180) * coordinate.Longitude;

            double N = radiusOfCurvature(coordinate.Latitude);

            Vector3 Result = new Vector3();
            Result.X = ((N + coordinate.Elevation ?? 0) * Math.Cos(radLat) * Math.Cos(radLon) + xShift);
            Result.Y = ((N + coordinate.Elevation ?? 0) * Math.Cos(radLat) * Math.Sin(radLon) + yShift);
            Result.Z = (((b * b * N) / (a * a)) * Math.Sin(radLat) + zShift);

            return Result;
        }

        public double Distance(Coordinate from, Coordinate to)
        {
            Vector3 diff = ToCarthesian(from) - ToCarthesian(to);
            return diff.Length();
        }

        public Coordinate FromCarthesian(Vector3 position)
        {
            Coordinate coord = new Coordinate();

            coord.Longitude = Math.Atan2(position.Y, position.X) * 180.0 / Math.PI;

            double p = Math.Sqrt((double)position.X * (double)position.X + (double)position.Y * (double)position.Y);
            double theta = Math.Atan2((double)position.Z * a, p * b);
            double phi = Math.Atan2((double) position.Z + e_ * e_ * b * Math.Pow(Math.Sin(theta), 3), p - e * e * a * Math.Pow(Math.Cos(theta), 3));

            coord.Latitude = phi * 180.0 / Math.PI;
            coord.Elevation =(float)(p / Math.Cos(phi) - radiusOfCurvature(coord.Latitude));

            return coord;
        }

        protected double radiusOfCurvature(double latitude)
        {
            return a / Math.Sqrt(1 - e * e * Math.Pow(Math.Sin(latitude * Math.PI / 180.0), 2));
        }
    }
}

File: web/webmap/src/geography/Datum.ts
import { DoubleVector3 } from "../geometry/DoubleVector3";
import { Coordinate } from "./Coordinate";

export class Datum {
    public static WGS84: Datum = new Datum(6378137, 1.0 / 298.257223563, 0, 0, 0);

    public static SmallDebug: Datum = new Datum(1000, 1.0 / 298.257223563, 0, 0, 0);

    protected a: number;
    protected b: number;
    protected e: number;
    protected e_: number;
    protected f: number;
    protected xShift: number;
    protected yShift: number;
    protected zShift: number;

    constructor(a: number, f: number, xShift: number, yShift: number, zShift: number) {
        this.a = a;
        this.f = f;
        this.b = a * (1 - f);
        this.xShift = xShift;
        this.yShift = yShift;
        this.zShift = zShift;

        this.e = Math.sqrt((a * a - this.b * this.b) / (a * a));
        this.e_ = Math.sqrt((a * a - this.b * this.b) / (this.b * this.b));
    }

    /**
     * Calculates the carthesian coordinates from a given coordinate.
     * The coordinate system is as follows:
     * 
     * z-axis goes through poles (+ is north, - south)
     * x-axis goes through 0/0 (0 is prime meridian, positive is east, negative is west)
     * y-axis is perpendicular to x and z, and goes through 0/90 (positive) and 0/-90 (negative)
     * @param coordinate Coordinate to convert
     * @returns Carthesian coordinate
     */
    public toCarthesian(coordinate: Coordinate): DoubleVector3 {
        const radLat = (Math.PI / 180) * coordinate.latitude;
        const radLon = (Math.PI / 180) * coordinate.longitude;

        const N = this.radiusOfCurvature(coordinate.latitude);

        const result = new DoubleVector3();
        const elevation = coordinate.elevation ?? 0;
        result.x = ((N + elevation) * Math.cos(radLat) * Math.cos(radLon) + this.xShift);
        result.y = ((N + elevation) * Math.cos(radLat) * Math.sin(radLon) + this.yShift);
        result.z = ((((this.b * this.b * N) / (this.a * this.a)) + elevation) * Math.sin(radLat) + this.zShift);

        return result;
    }

    /**
     * Calculates the distance between two coordinates. Beware that this is a straight 
     * line distance, not a distance along the surface.
     * @param from First coordinate
     * @param to Second coordinate
     * @returns Distance between the two coordinates
     */
    public distance(from: Coordinate, to: Coordinate): number {
        const diff = this.toCarthesian(from).subtract(this.toCarthesian(to));
        return diff.length();
    }

    /**
     * Converts a Cartesian coordinate to a geographic coordinate (latitude, longitude, elevation).
     * 
     * @param pos - The Cartesian coordinate to convert.
     * @returns The geographic coordinate corresponding to the given Cartesian coordinate.
     */
    public fromCarthesian(pos: DoubleVector3): Coordinate {
        const coord = new Coordinate();

        const position = new DoubleVector3(pos.x - this.xShift, pos.y - this.yShift, pos.z - this.zShift);

        // Calculate longitude
        coord.longitude = Math.atan2(position.y, position.x) * 180.0 / Math.PI;

        // Calculate intermediate values
        const p = Math.sqrt(position.x * position.x + position.y * position.y);
        const theta = Math.atan2(position.z * this.a, p * this.b);
        const phi = Math.atan2(position.z + this.e_ * this.e_ * this.b * Math.pow(Math.sin(theta), 3), p - this.e * this.e * this.a * Math.pow(Math.cos(theta), 3));

        // Calculate latitude
        coord.latitude = phi * 180.0 / Math.PI;

        // Calculate elevation
        const N = this.radiusOfCurvature(coord.latitude);
        coord.elevation = p / Math.cos(phi) - N;

        return coord;
    }
    
    protected radiusOfCurvature(latitude: number): number {
        return this.a / Math.sqrt(1 - this.e * this.e * Math.pow(Math.sin(latitude * Math.PI / 180.0), 2));
    }

    public get meridianLength() {
        return Math.PI * 2 * this.a;
    }
}
